### PR TITLE
[11.0][MIG] web_tree dynamic_colored_field

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: web_tree_dynamic_colored_field/README.rst [hunk 1]
```diff
@@ -0,0 +1,105 @@
+Colorize field in tree views
```
Comment (simahawk) [inline web_tree_dynamic_colored_field/README.rst:6]: update readme according to https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst

File: web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+            this.colorField = this.arch.attrs.colors.split(';')
+                .filter(color => color.startsWith('color_field'))[0]
+                .split(':')[1]
+                .trim();
+            if (this.colorField) {
+                // validate the presence of that field in tree view
+                var fieldNames = _(this.columns).map(
+                    (value) => { return value.attrs.name; }
+                );
+                if (fieldNames.indexOf(this.colorField) === -1) {
```
Comment (simahawk) [inline web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js]: I'd say you should just log here. If you keep it you should use `_t` to translate the msg.

File: web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+        /**
+         * Colorize the current cell depending on expressions provided.
+         *
+         * @param {Query Node} $td a <td> tag inside a table representing a list view
+         * @param {Object} node an XML node (must be a <field>)
+         */
+        applyColorize: function ($td, node, ctx) {
+            if (!node.attrs.options) { return; }
+            var nodeOptions = JSON.parse(node.attrs.options);
```
Comment (simahawk) [inline web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js]: don't you get already parse options here?

File: web_tree_dynamic_colored_field/README.rst [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+* Add attribute ``bg_color`` on field's ``options`` to color background of a cell in tree view
+
+* Add attribute ``fg_color`` on field's ``options`` to change text color of a cell in tree view
+
+* Add attribute ``color_field`` on the tree element's ``options`` to use as color
+
+
+Usage
+=====
```
Comment (simahawk) [inline web_tree_dynamic_colored_field/README.rst]: is this up to date?

File: web_tree_dynamic_colored_field/README.rst [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    
+    With this example, column which renders 'name' field will have its text colored in white.
+
+* In the tree view declaration, use ``options='"color_field": "color"'`` attribute in the ``tree`` tag::
+
+    ...
+    <field name="arch" type="xml">
+        <tree string="View name" colors="color_field: my_color" >
+            ...
+            <field name="color" invisible="1" />
```
Comment (simahawk) [inline web_tree_dynamic_colored_field/README.rst]: color/my_color

**Review by simahawk — CHANGES_REQUESTED** [review]

File: web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js [hunk 6]
```diff
@@ -1,95 +1,119 @@
...
-                    expression = pair[1];
-                if (py.evaluate(expression, ctx).toJSON()) {
-                    result = css_attribute + ': ' + color + ';';
+    ListRenderer.include({
+        /**
+         * @override
+         */
+        _renderBody: function () {
+            this.colorField = this.arch.attrs.colors.split(';')
+                .filter(color => color.startsWith('color_field'))[0]
```
Comment (ghost) [inline web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js]: should `trim()` the `color` variable before performing a check

File: web_tree_dynamic_colored_field/README.rst [hunk 7]
```diff
@@ -4,58 +4,58 @@ Colorize field in tree views
 This module aims to add support for dynamically coloring fields in tree view
 according to data in the record.
 
-It provides attributes on fields with the same syntax as the 'colors' attribute
+It provides attributes on fields with the same syntax as the ``colors`` attribute
```
Comment (ghost) [inline web_tree_dynamic_colored_field/README.rst]: actually, those are not quite same

File: web_tree_dynamic_colored_field/README.rst [hunk 8]
```diff
@@ -4,58 +4,58 @@ Colorize field in tree views
...
 
-* In the tree view declaration, put bg_color="red:customer==True;" attribute in the field tag::
+* In the tree view declaration, put ``options='"bg_color": "red: customer==True"`` attribute in the ``field`` tag::
 
     ...
     <field name="arch" type="xml">
         <tree string="View name">
             ...
-            <field name="name" bg_color="red:customer==True;"/>
+            <field name="name" options='"bg_color": "customer == True"'/>
```
Comment (ghost) [inline web_tree_dynamic_colored_field/README.rst]: misses a color

File: web_tree_dynamic_colored_field/README.rst [hunk 9]
```diff
@@ -4,58 +4,58 @@ Colorize field in tree views
...
+            <field name="name" 'options="fg_color": "white:customer == True"'/>
             ...
         </tree>
     </field>
     ...
     
     With this example, column which renders 'name' field will have its text colored in white.
 
-* In the tree view declaration, use color_field="color" attribute in the tree tag::
+* In the tree view declaration, use ``options='"color_field": "color"'`` attribute in the ``tree`` tag::
```
Comment (ghost) [inline web_tree_dynamic_colored_field/README.rst]: s/color/my_color

File: web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js [hunk 10]
```diff
@@ -1,95 +1,119 @@
...
+                .split(':')[1]
+                .trim();
+            if (this.colorField) {
+                // validate the presence of that field in tree view
+                var fieldNames = _(this.columns).map(
+                    (value) => { return value.attrs.name; }
+                );
+                if (fieldNames.indexOf(this.colorField) === -1) {
+                    // this could be replaced by RPC call tho...
```
Comment (ghost) [inline web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js]: ...but not here, not inside of a `_renderBody`

File: web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js [hunk 11]
```diff
@@ -1,95 +1,119 @@
...
+         * Colorize the current cell depending on expressions provided.
+         *
+         * @param {Query Node} $td a <td> tag inside a table representing a list view
+         * @param {Object} node an XML node (must be a <field>)
+         */
+        applyColorize: function ($td, node, ctx) {
+            if (!node.attrs.options) { return; }
+            var nodeOptions = JSON.parse(node.attrs.options);
+            if (this.colorField) {
+                $td.css('color', ctx[this.colorField]);
```
Comment (ghost) [inline web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js]: check for empty values

**Review by gurneyalex — APPROVED** [review]
tested works fine.

**Comment by ghost** [comment]
@simahawk @arkostyuk adressed :smile:

File: web_tree_dynamic_colored_field/views/web_tree_dynamic_colored_field.xml [hunk 12]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- vim:fdn=3:
```
Comment (simahawk) [inline web_tree_dynamic_colored_field/views/web_tree_dynamic_colored_field.xml]: drop this

File: web_tree_dynamic_colored_field/i18n/es.po [hunk 13]
```diff
@@ -0,0 +1,26 @@
+# Translation of Odoo Server.
```
Comment (simahawk) [inline web_tree_dynamic_colored_field/i18n/es.po]: I think you can trash all the translations: they are not needed anymore

File: web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+                var colorField = this.arch.attrs.colors.split(';')
+                .filter(color => color.trim().startsWith('color_field'))[0]
+                .split(':')[1]
+                .trim();
+                // validate the presence of that field in tree view
+                var fieldNames = _(this.columns).map(
+                    (value) => { return value.attrs.name; }
+                );
+                if (fieldNames.indexOf(colorField) === -1) {
+                    // this could be replaced by RPC call at a cell render tho...
```
Comment (simahawk) [inline web_tree_dynamic_colored_field/static/src/js/web_tree_dynamic_colored_field.js]: drop this comment

**Review by simahawk — CHANGES_REQUESTED** [review]

**Comment by simahawk** [comment]
@hbrunn would you mind doing a review loop here? :wink:

File: web_tree_dynamic_colored_field/views/web_tree_dynamic_colored_field.xml (hunk 12 above)
Comment (ghost) [reply web_tree_dynamic_colored_field/views/web_tree_dynamic_colored_field.xml]: thanks, comletely forgot about it

**Review by tschanzt — APPROVED** [review]
looks good to me.

**Review by simahawk — APPROVED** [review]

**Review by max3903 — APPROVED** [review]

**Comment by sudhir-erpharbor** [comment]
Please check this issue: #1103

**Pull request merged**